### PR TITLE
Postpone PV deletion with finalizer when it is being used

### LINKED ISSUE #33355: Use finalizer to delete-inhibit PVs while bound to a PVC
https://github.com/kubernetes/kubernetes/issues/29033#issuecomment-241055560

If the storage controller put a finalizer on the PV, then we could be relatively assured that nobody will delete the PV while a PVC is bound to it.

@kubernetes/sig-storage

### PR DESCRIPTION
**Pull request by NickrenREN**
Postpone PV deletion if it is bound to a PVC

xref: https://github.com/kubernetes/community/pull/1608

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #33355

**Special notes for your reviewer**:

**Release note**:
```release-note
Postpone PV deletion when it is being bound to a PVC
```

WIP, assign to myself first

/assign @NickrenREN

### CODE REVIEW

File: pkg/controller/volume/pvprotection/pv_protection_controller.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+}
+
+func (c *PVProtectionController) isBeingUsed(pv *v1.PersistentVolume) bool {
+	// check if PV is being bound to a PVC by its status
+	// the status will be updated by PV controller
+	if pv.Status.Phase == v1.VolumeFailed || pv.Status.Phase == v1.VolumeReleased {
+		// the PV is not being used now
+		return false
+	}
+	if pv.Status.Phase == v1.VolumeAvailable && pv.Spec.ClaimRef == nil {
```
Comment (jsafrane) [inline pkg/controller/volume/pvprotection/pv_protection_controller.go]: Admin can post a pre-bound PV (`pv.Spec.ClaimRef` is not nil) and they should be allowed to delete it until user does not create appropriate PVC. IMO, we could trust `pv.Status.Phase==VolumeAvailable` here without races - PV controller won't bind anything that's marked for deletion.

Please allow deletion of `Pending` volumes too.

Comment (jsafrane) [reply pkg/controller/volume/pvprotection/pv_protection_controller.go]: Now that I think about it, can we leave all the dirty work to PV controller and ignore PVCs completely in this PR? PV controller makes sure that PVs are Released or Failed when they're ready to be deleted...

This protection controller would only ensure that PV that's Bound can't be deleted.

Comment (NickrenREN) [reply pkg/controller/volume/pvprotection/pv_protection_controller.go]: Another situation:  a PV is bound to a PVC, but the PVC has deletionTimeStamp set, should we allow admin to delete the PV ?  Or wait util PVC is deleted and PV is not bound ?

Comment (jsafrane) [reply pkg/controller/volume/pvprotection/pv_protection_controller.go]: > a PV is bound to a PVC, but the PVC has deletionTimeStamp set, should we allow admin to delete the PV

No, the PVC is needed by something that holds the finalizer, e.g. a running pod. We must not delete the PV while it's used by a pod!

File: cmd/kube-controller-manager/app/core.go [hunk 2]
```diff
@@ -403,3 +404,15 @@ func startPVCProtectionController(ctx ControllerContext) (bool, error) {
 	}
 	return false, nil
 }
+
+func startPVProtectionController(ctx ControllerContext) (bool, error) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.StorageProtection) {
+		go pvprotection.NewPVProtectionController(
+			ctx.InformerFactory.Core().V1().PersistentVolumes(),
+			ctx.InformerFactory.Core().V1().PersistentVolumeClaims(),
+			ctx.ClientBuilder.ClientOrDie("pv-protection-controller"),
```
Comment (jsafrane) [inline cmd/kube-controller-manager/app/core.go]: I think you need to add a policy to `pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go`.

File: pkg/controller/volume/pvprotection/pv_protection_controller.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+// PVProtectionController is controller that removes PVProtectionFinalizer
+// from PVs that are not bound to PVCs.
+type PVProtectionController struct {
+	client clientset.Interface
+
+	pvLister       corelisters.PersistentVolumeLister
+	pvListerSynced cache.InformerSynced
+
+	pvcLister       corelisters.PersistentVolumeClaimLister
```
Comment (jsafrane) [inline pkg/controller/volume/pvprotection/pv_protection_controller.go]: is PVC lister still necessary?

File: pkg/controller/volume/pvprotection/pv_protection_controller.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+// pvcAddedDeletedUpdated reacts to PVC events
+func (c *PVProtectionController) pvcAddedDeletedUpdated(obj interface{}, deleted bool) {
+	pvc, ok := obj.(*v1.PersistentVolumeClaim)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("PVC informer returned non-PVC object: %#v", obj))
+		return
+	}
+
+	// Filter out pvc that can't help us to remove a finalizer on PV
+	if !deleted && pvc.ObjectMeta.DeletionTimestamp == nil {
```
Comment (jsafrane) [inline pkg/controller/volume/pvprotection/pv_protection_controller.go]: as you asked above, we should not delete PVs that are bound to PVCs with deletion timestamp, so checking `DeletionTimestamp` is probably useless.

Correct me if I am wrong, I think that PV controller will mark a PV as Released when a PVC is deleted so this protection controller might not need to watch PVCs at all and rely on PV controller to do the checks. Is there a case where the check above would help?

File: cmd/kube-controller-manager/app/core.go (hunk 2 above)
Comment (NickrenREN) [reply cmd/kube-controller-manager/app/core.go]: yeah, will add when finished the implementation

File: pkg/controller/volume/pvprotection/pv_protection_controller.go (hunk 4 above)
Comment (NickrenREN) [reply pkg/controller/volume/pvprotection/pv_protection_controller.go]: If we are not allowed to delete PVs that are bound to PVCs with deletion timestamp,  i think you are right, PVC sruff here is useless

Comment (NickrenREN) [reply pkg/controller/volume/pvprotection/pv_protection_controller.go]: If we just need to watch PV events here, maybe can consider merging it with PVC protection controller.

File: pkg/controller/volume/pvprotection/pv_protection_controller.go (hunk 3 above)
Comment (NickrenREN) [reply pkg/controller/volume/pvprotection/pv_protection_controller.go]: I am afraid not

File: hack/.golint_failures [hunk 5]
```diff
@@ -130,6 +130,7 @@ pkg/controller/volume/events
 pkg/controller/volume/expand
 pkg/controller/volume/persistentvolume
 pkg/controller/volume/persistentvolume/options
+pkg/controller/volume/pvprotection
```
Comment (jsafrane) [inline hack/.golint_failures]: I'd appreciate if the PR could be without this line. I guess it complains about type/function names and a little rename would resolve it.

File: pkg/controller/volume/pvprotection/pv_protection_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (jsafrane) [inline pkg/controller/volume/pvprotection/pv_protection_controller_test.go]: 2018 :-D

**Review by jsafrane — COMMENTED** [review]
I have just two minor nits, otherwise LGTM. And thanks for really good test coverage!

File: pkg/controller/volume/pvprotection/pv_protection_controller_test.go (hunk 6 above)
Comment (NickrenREN) [reply pkg/controller/volume/pvprotection/pv_protection_controller_test.go]: thanks, updated

File: hack/.golint_failures (hunk 5 above)
Comment (NickrenREN) [reply hack/.golint_failures]: thanks,done

**Comment by NickrenREN** [comment]
the e2e-gke test error is not related to this PR and it seems has no effect

**Review by zouyee — APPROVED** [review]

File: pkg/features/kube_features.go [hunk 7]
```diff
@@ -203,10 +203,10 @@ const (
 	BlockVolume utilfeature.Feature = "BlockVolume"
 
 	// owner: @pospispa
-	//
 	// alpha: v1.9
-	// Postpone deletion of a persistent volume claim in case it is used by a pod
-	PVCProtection utilfeature.Feature = "PVCProtection"
+	//
+	// Postpone deletion of a PV or a PVC when they are being used
+	StorageProtection utilfeature.Feature = "StorageProtection"
```
Comment (thockin) [inline pkg/features/kube_features.go]: This makes it sound like something else entirely.  Can we rename this to "PVDeleteInhibit" or "VolumeInUseProtection" or something?

File: cluster/centos/config-default.sh [hunk 8]
```diff
@@ -120,7 +120,7 @@ export FLANNEL_NET=${FLANNEL_NET:-"172.16.0.0/16"}
 
 # Admission Controllers to invoke prior to persisting objects in cluster
 # If we included ResourceQuota, we should keep it at the end of the list to prevent incrementing quota usage prematurely.
-export ADMISSION_CONTROL=${ADMISSION_CONTROL:-"Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeClaimResize,DefaultTolerationSeconds,Priority,PVCProtection,ResourceQuota"}
```
Comment (thockin) [inline cluster/centos/config-default.sh]: This should be in the first commit

File: cluster/gce/config-default.sh [hunk 9]
```diff
@@ -298,7 +298,7 @@ if [[ -n "${GCE_GLBC_IMAGE:-}" ]]; then
 fi
 
 # Admission Controllers to invoke prior to persisting objects in cluster
-ADMISSION_CONTROL=Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,PersistentVolumeClaimResize,DefaultTolerationSeconds,NodeRestriction,Priority,PVCProtection
```
Comment (thockin) [inline cluster/gce/config-default.sh]: This should be in the first commit

File: pkg/features/kube_features.go (hunk 7 above)
Comment (NickrenREN) [reply pkg/features/kube_features.go]: we did not pay much attention on the name. in order to reuse this feature gate for PV(it is introduced for PVC), maybe later also for SC, we change it to StorageProtection . 
"PVDeleteInhibit" or "VolumeInUseProtection" sounds like just for PV ?

File: cluster/centos/config-default.sh (hunk 8 above)
Comment (NickrenREN) [reply cluster/centos/config-default.sh]: It (PVCProtection) was introduced before. this PR reuse it for PV Protection and change the name in the first commit

File: pkg/features/kube_features.go (hunk 7 above)
Comment (jsafrane) [reply pkg/features/kube_features.go]: I opened https://github.com/kubernetes/kubernetes/issues/59639

**Pull request merged**